src/irmin/inode.ml

```diff
@@ -0,0 +1,983 @@
+                  { length = 0; depth; entries = Array.make Conf.entries None }
+              in
+              let aux t (s, v) =
+                (add [@tailcall]) layout ~depth ~copy:false ~replace t s v
+                  (fun x -> x)
+              in
+              List.fold_left aux empty vs
+          in
+          k t
+      | Tree t -> (
+          let length = if replace then t.length else t.length + 1 in
+          let entries = if copy then Array.copy t.entries else t.entries in
+          let i = index ~depth s in
+          match entries.(i) with
+          | None ->
+              let target = values layout (StepMap.singleton s v) in
+              entries.(i) <- Some (Ptr.of_target layout target);
+              let t = tree layout { depth; length; entries } in
+              k t
+          | Some n ->
+              let t = Ptr.target layout n in
+              add layout ~depth:(depth + 1) ~copy ~replace t s v
+              @@ fun target ->
+              entries.(i) <- Some (Ptr.of_target layout target);
+              let t = tree layout { depth; length; entries } in
+              k t)
+
+    let add layout ~copy t s v =
+      (* XXX: [find_value ~depth:42] should break the unit tests. It doesn't. *)
+      match find_value ~depth:0 layout t s with
+      | Some v' when equal_value v v' -> stabilize layout t
+      | Some _ ->
+          add ~depth:0 layout ~copy ~replace:true t s v Fun.id
+          |> stabilize layout
+      | None ->
+          add ~depth:0 layout ~copy ~replace:false t s v Fun.id
+          |> stabilize layout
+
+    let rec remove layout ~depth t s k =
+      match t.v with
+      | Values vs ->
+          let t = values layout (StepMap.remove s vs) in
+          k t
+      | Tree t -> (
+          let len = t.length - 1 in
+          if len <= Conf.entries then
+            let vs =
+              list_tree layout ~offset:0 ~length:t.length (empty_acc t.length) t
+            in
+            let vs = List.concat (List.rev vs.values) in
+            let vs = StepMap.of_list vs in
+            let vs = StepMap.remove s vs in
+            let t = values layout vs in
+            k t
+          else
+            let entries = Array.copy t.entries in
+            let i = index ~depth s in
+            match entries.(i) with
+            | None -> assert false
+            | Some t ->
+                let t = Ptr.target layout t in
+                if length t = 1 then (
+                  entries.(i) <- None;
+                  let t = tree layout { depth; length = len; entries } in
+                  k t)
+                else
+                  remove ~depth:(depth + 1) layout t s @@ fun target ->
+                  entries.(i) <- Some (Ptr.of_target layout target);
+                  let t = tree layout { depth; length = len; entries } in
+                  k t)
+
+    let remove layout t s =
+      (* XXX: [find_value ~depth:42] should break the unit tests. It doesn't. *)
```

Comment:
same here: is `XXX` still up-to-date? Doesn't seem to match with the function name.

Reply:
I think so, `find_value` refers to the call on the next line